repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py

```diff
@@ -0,0 +1,339 @@
+        path = '/etc/cups' + path
+      # move certificate or key to supported default path - /etc/cups/ssl - if it is not
+      # already
+      if not path.startswith('/etc/cups/ssl'):
+        os.rename(path, '/etc/cups/ssl/' + path.rsplit('/', 1)[1])
+        # if file was moved to default directory, no need to keep directive 
+        # in configuration file
+        new_content = self.remove_string(new_directive, path, new_content)
+      else:
+        if os.path.isfile(path):
+          # if path starts with /etc/cups/ssl, remove the filename from the path,
+          # because ServerKeychain wants directory
+          new_path = path.rsplit('/', 1)[0]
+          new_content = new_content.replace(path, new_path)
+
+
+      return new_content
+
+    def find_values(self, directive, content):
+      """
+      migrate_serverkeychain
+
+      parameters:
+      - directive - string
+      - new_content - string
+
+      returns list of strings
+      """
+
+      # we need to get values
+      value_regexp = '^\s*' + directive + '\s+(.+?)\s' 
+      value_pattern = compile(value_regexp, MULTILINE)
+      found_values = findall(value_pattern, content)
+
+      return found_values
+
+    def find_and_replace(self, directives, new_directive, file):
+      """
+      find_and_replace
+  
+      parameters:
+      - directives - list of strings
+      - new-directive - string
+      - file - string
+  
+      returns boolean
+      """
+
+      result = False
+
+      for directive in directives:
+        # regexp and new string differs if we replace directive or value - directive needs to have at least one
+        # white character at the end, but value needs to have one white character at start and zero or N white
+        # characters at the end + new line character
+        if directive in changed_directives:
+          directive_regexp = directive + '\s'
+          new_directive_string = new_directive + ' '
+        elif directive in changed_values:
+          directive_regexp = '\s' + directive + '\s*\n'
+          new_directive_string = ' ' + new_directive + '\n'
+        pattern = compile(directive_regexp)
+
+        try:
+          with open(file, 'r') as conf:
+            content = conf.read()
+        except IOError:
+          return result
+
+        new_content = sub(pattern, new_directive_string, content)
+        if new_directive == 'ServerKeychain':
+          # ServerKeyChain migration needs more attention - like moving the files into specific directory
+          found_values = self.find_values(new_directive, new_content)
+          for value in found_values:
+            new_content = self.migrate_serverkeychain_path(value, new_directive, new_content)
+
+        try:
+          with open(file, 'w') as conf:
+            conf.write(new_content)
+          result = True
+        except IOError:
+          return result
+
+      return result
+
+    def migrate_includes(self, include_files):
```

Comment:
Same thing for this function, instead of returning boolean maybe we could return a descriptive error if something happened or empty string if everything is OK. Also, maybe would be a good practice to check if all files exists and are readable/writeable before doing proper changes. If the process fails in the end, but some changes were applied we can end up with a not working system.

Reply:
Now I check cupsd.conf and cups-files.conf for availability before doing anything else, exiting with error message if CUPS is installed but confs are missing.
Ad included files - their absence does not crash CUPS, just adds error into log - IMO it is user error to put non-existing file into configuration. It generates error (but service is still running) even on RHEL 7, so I chose to let 'Include' directives with non-existing files in cupsd.conf.